Mkirk/fix tests

michaelkirk opened the pull request:
PTAL @charlesmchen 

I think it's all non-controversial, except maybe:
https://github.com/WhisperSystems/SignalServiceKit/commit/819f3f1eb98910625ae15a420e3009bf5444bc9c

michaelkirk committed:
Fixed test build. Some tests still failing.

    Executed 85 tests, with 22 failures (17 unexpected) in 7.416 (8.531) seconds

// FREEBIE

michaelkirk committed:
fix disappearing messages test

// FREEBIE

michaelkirk committed:
Fix up some more tests.

// FREEBIE

michaelkirk committed:
set-expiration dispatches *sync* so we can test it.

// FREEBIE

michaelkirk committed:
Allow override of singleton enforcement in test app

We frequently build multiple instances of "singletons" for the purpose
of testing.

// FREEBIE

charlesmchen commented on src/Messages/OWSDisappearingMessagesJob.m:
```diff
@@ -121,7 +121,7 @@ - (void)runLoop
 + (void)setExpirationForMessage:(TSMessage *)message
 {
-    dispatch_async(self.serialQueue, ^{
+    dispatch_sync(self.serialQueue, ^{
```
I don't think we want to do this. This queue can get a bit backed up. If we need to make this testable, we could just have the test wait, or add a completion, etc.

charlesmchen replied:
@michaelkirk Did you see this comment?

michaelkirk replied:
Yup, that's why it's not merged. 

I think I'd prefer to add a completion handler vs. getting in the habit of adding sleep(1) in tests, but that would take more than a minute of dev time. 

This PR hasn't been top priority for me, so feel free to run with it if you're feeling frisky.

michaelkirk replied:
PTAL @charlesmchen 

I considered adding a completion handler, but the method chain is really deep, across multiple dispatch_asyncs. So I opted for a `sleep` hack in the tests. I really don't like to do this, but I think it's more important we get the tests running.

e.g. since I last picked up this pull request there were 3 or so other changes that broke a bunch of tests. Commits to fix those are also included in this PR.

charlesmchen commented on Example/TSKitiOSTestApp/Podfile:
```diff
@@ -11,3 +11,18 @@ target 'TSKitiOSTestApp' do
+          existing_definitions = "$(inheritied)"
+        end
+
+        config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] = "#{existing_definitions} SSK_NO_ENFORCE_SINGLETONS=1"
```
Instead of SSK_NO_ENFORCE_SINGLETONS how about BUILDING_FOR_TESTS or something generic like that?

michaelkirk replied:
SSK_BUILDING_FOR_TESTS works for me.

michaelkirk committed:
CR: renaming var to be more general

// FREEBIE

michaelkirk committed:
prefer hacks in test to hacks in code.

Adding a completion handler would be a bit gnarly since we're snaking
through a pretty long method chain.

// FREEBIE

michaelkirk committed:
Fix some more tests.

Adapt to IdentityManager refactor
Adapt to DisappearingMessages refactor

// FREEBIE

michaelkirk committed:
Fix up deleteAttachments method since making attachmentFolder
dispatchOnce

// FREEBIE

charlesmchen reviewed (approved):
LGTM

michaelkirk committed:
gemfile for consistent build on dev/CI

// FREEBIE

michaelkirk committed:
bump travis image to 8.3

// FREEBIE

michaelkirk committed:
try fastlane scan, since build is timing out.

// FREEBIE
